### PR TITLE
Fix newline rendering

### PR DESCRIPTION
**Pull request by LightAndLight**
Due to the relatively unstructure way newlines are stored, there's a lot of potential to have missing newlines in places where newlines are required. I'd like for things to be more structurally correct, but I tried really hard and couldn't make it happen. In the meantime, I've modified the render to correct the pathological cases, because I think it's too tedious to be reported as a syntax error.

I'm mainly wondering if I should completely factor out that correction behaviour, rather than having it built in to the renderer.

### CODE REVIEW

**Comment by gwils** [comment]
Are there cases where we want the newline correction to happen outside rendering? (would it help derived packages like my reindent thing?)

If yes, then I think factoring it out is a worthwhile thing. How much effort is it? Is it something that we should make a note of and do later?

**Comment by LightAndLight** [comment]
I'm not sure if newline correction is needed outside rendering. ASTs produced by the parser or written in the DSL always have good newlines. Also, newlines don't have any semantic value in the AST. 

It's mainly a problem that is theoretically possible, but practically unlikely

File: src/Language/Python/Internal/Render.hs [hunk 1]
```diff
@@ -49,57 +55,66 @@ import Language.Python.Internal.Token (PyToken(..))
...
 
 -- | A 'RenderOutput' is an intermediate form used during rendering
--- with efficient 'Semigroup' and 'Monoid' instances.
-newtype RenderOutput
+-- with efficient concatenation
+newtype RenderOutput a
   = RenderOutput
-  { unRenderOutput :: DList (PyToken ())
-  } deriving (Eq, Show, Semigroup, Monoid)
+  { unRenderOutput :: ReaderT Bool (Writer (DList (PyToken ()))) a
```
[the remark this answers is not in the document]
Comment (LightAndLight) [reply src/Language/Python/Internal/Render.hs]: Yeah I'll do this

[the remark this answers is not in the document]
Comment (LightAndLight) [reply src/Language/Python/Internal/Render.hs]: oh wait actually I do use the `a`, e.g. in `isFinal`http://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Monoid.html#t:Ap

[the remark this answers is not in the document]
Comment (gwils) [reply src/Language/Python/Internal/Render.hs]: Sweet. In that case, I recommend the change I suggested to bracket. As a broader comment, I think `RenderOutput ()` is required in a lot of places where we could have `RenderOutput a`.

**Comment by gwils** [comment]
In that case, I say don't bother splitting it out yet.

**Pull request merged**